[Flaking Test] [sig-auth] integration-master	TestMultiWebhookAuthzConfig

### Which jobs are flaking?

- ci-kubernetes-integration-master

### Which tests are flaking?

--- FAIL: TestMultiWebhookAuthzConfig (9.23s)
k8s.io/kubernetes/test/integration/auth.auth


### Since when has it been flaking?

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1770274294463467520
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1773485209702895616 (03-20)

I cannot search for previous flake. 

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#integration-master

### Reason for failure (if possible)

```
    authz_config_test.go:669: failure timestamp changed from reload1Metrics.reloadFailure 2024-03-28 23:27:10.466200351 +0000 UTC to 2024-03-28 23:27:10.953506469 +0000 UTC unexpectedly
```

### Anything else we need to know?

https://github.com/kubernetes/kubernetes/blame/f4e246bc93ffb68b33ed67c7896c379efa4207e7/test/integration/auth/authz_config_test.go#L124 
updated recently

### Relevant SIG(s)

/sig auth
/cc @liggitt

--- a/test/integration/auth/authz_config_test.go
+++ b/test/integration/auth/authz_config_test.go
@@ -56,7 +56,7 @@ func TestAuthzConfig(t *testing.T) {
 
 	dir := t.TempDir()
 	configFileName := filepath.Join(dir, "config.yaml")
-	if err := os.WriteFile(configFileName, []byte(`
+	if err := atomicWriteFile(configFileName, []byte(`
 apiVersion: apiserver.config.k8s.io/v1alpha1
 kind: AuthorizationConfiguration
 authorizers:
@@ -362,7 +362,7 @@ users:
 	}
 
 	configFileName := filepath.Join(dir, "config.yaml")
-	if err := os.WriteFile(configFileName, []byte(`
+	if err := atomicWriteFile(configFileName, []byte(`
 apiVersion: apiserver.config.k8s.io/v1alpha1
 kind: AuthorizationConfiguration
 authorizers:
@@ -583,7 +583,7 @@ authorizers:
 	}
 
 	// write bogus file
-	if err := os.WriteFile(configFileName, []byte(`apiVersion: apiserver.config.k8s.io`), os.FileMode(0644)); err != nil {
+	if err := atomicWriteFile(configFileName, []byte(`apiVersion: apiserver.config.k8s.io`), os.FileMode(0644)); err != nil {
 		t.Fatal(err)
 	}
 
@@ -635,7 +635,7 @@ authorizers:
 	}
 
 	// write good config with different webhook
-	if err := os.WriteFile(configFileName, []byte(`
+	if err := atomicWriteFile(configFileName, []byte(`
 apiVersion: apiserver.config.k8s.io/v1beta1
 kind: AuthorizationConfiguration
 authorizers:
@@ -889,4 +889,12 @@ func getMetrics(t *testing.T, client *clientset.Clientset) (*metrics, error) {
 		}
 	}
 	return &m, nil
+}
+
+func atomicWriteFile(name string, data []byte, perm os.FileMode) error {
+	tmp := name + ".tmp"
+	if err := os.WriteFile(tmp, data, perm); err != nil {
+		return err
+	}
+	return os.Rename(tmp, name)
 }

